File: Year 1/Semester 2/Java/Lessons/overloading_and_obj_compr/Student_overload.java
package overloading_and_obj_compr;

public class Student_overload implements Comparable<Student> {
    // The Comparable interface allows us to compare two objects

    private String name;
    
    public Student_overload(String name) {
        this.name = name;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null || getClass() != obj.getClass()) {
            return false;
        }
        Student_overload other = (Student_overload) obj;
        return this.name.equals(other.name);
    }

    @Override
    public int compareTo(Student other) {
        return this.name.compareTo(other.name);
    }

    // Getter for name
    public String getName() {
        return name;
    }

    // Setter for name
    public void setName(String name) {
        this.name = name;
    }

    public String toString() {
        return "Student(" + this.name +")";
    }


}

File: Year 1/Semester 1/C/Lessons/overloading_and_obj_compr/Student_overload.c
#include <stdio.h>
#include <string.h>

// Define the Student struct
typedef struct {
    char name[50];
} Student;

// Constructor function for Student
Student createStudent(const char *name) {
    Student s;
    strcpy(s.name, name);
    return s;
}

// Equality check function (equivalent to `equals` method in Java)
int isEqual(Student s1, Student s2) {
    return strcmp(s1.name, s2.name) == 0;
}

// Comparison function (equivalent to `compareTo` method in Java)
int compareTo(Student s1, Student s2) {
    return strcmp(s1.name, s2.name);
}

// Getter for the name field
const char* getName(Student s) {
    return s.name;
}

// Setter for the name field
void setName(Student *s, const char *name) {
    strcpy(s->name, name);
}

// ToString equivalent function for displaying the student
void toString(Student s) {
    printf("Student(%s)\n", s.name);
}

int main() {
    // Creating two Student objects
    Student student1 = createStudent("Alice");
    Student student2 = createStudent("Bob");

    // Displaying the students
    toString(student1);
    toString(student2);

    // Comparing students
    if (isEqual(student1, student2)) {
        printf("The students are equal.\n");
    } else {
        printf("The students are not equal.\n");
    }

    // Using the comparison function
    int comparison = compareTo(student1, student2);
    if (comparison < 0) {
        printf("%s comes before %s alphabetically.\n", student1.name, student2.name);
    } else if (comparison > 0) {
        printf("%s comes after %s alphabetically.\n", student1.name, student2.name);
    } else {
        printf("The students have the same name.\n");
    }

    return 0;
}


